### PR TITLE
Pytorch Finetuning

### PR DESCRIPTION
**Pull request by jjbrophy47**
Hi, I really like this repository and how easy it use to use as a PyTorch alternative for GPT2. In this pull request, I've added the ability to fine-tune a pre-trained gpt2 model in PyTorch. I've adapted training code from nshepperd: https://github.com/nshepperd/gpt-2/blob/finetuning/train.py

I hope you find this useful! Let me know if you have any questions or concerns!
-Jonathan Brophy

### CODE REVIEW

**Comment by graykode** [comment]
Hello. First, Thank you for your sincere pull request!
I will read your code in line by line as much as possible. 👍

**Comment by graykode** [comment]
I read your code quickly and I have some questions.
 I left code review comment on line corresponding to the question.

File: train.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    else:
+        print('model does not exist at: {}'.format(model_path))
+        exit(0)
+    return state_dict
+
+
+def load_model(model, state_dict, device):
+    model = load_weight(model, state_dict)
+    model.to(device)
+    model.eval()
```
Comment (graykode) [inline train.py]: @jjbrophy47  Why did you try `model.eval()`?
Didn't you mean fine tuning?

File: GPT2/data.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        if path.endswith('.npz'):
+
+            # Pre-encoded
+            with np.load(path) as npz:
+                for item in npz.files:
+                    token_chunks.append(npz[item])
+        else:
+
+            # Plain text
+            with open(path, 'r') as fp:
```
Comment (graykode) [inline GPT2/data.py]:
```python
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 43: ordinal not in range(128)
```
UnicodeDecodeError is occured in here. I think open with `utf-8` is right.

File: train.py (hunk 1 above)
Comment (jjbrophy47) [reply train.py]: Ooops! That line is a mistake and should be deleted, thanks.

File: GPT2/data.py (hunk 2 above)
Comment (jjbrophy47) [reply GPT2/data.py]: Ok I think that would be better, thanks.

**Comment by graykode** [comment]
branch name `archive` is branch to record before commits, so could you repull request to `train` branch? Thank you again.

**Comment by graykode** [comment]
Thanks! I will also reflect the modifications to the master branch.

**Pull request merged**